lib/index.js

```diff
@@ -0,0 +1,169 @@
+  if (!fs.existsSync(`${OUTPUT_FOLDER}/${dayString}/${hourString}`)) {
+    fs.mkdirSync(`${OUTPUT_FOLDER}/${dayString}/${hourString}`);
+  }
+
+  return { dayString, hourString };
+};
+
+function saveToFile(summary) {
+  summary = summary.trim();
+
+  const { dayString, hourString } = fileSystemSetup();
+
+  const rand = Math.floor(Math.random() * 1000000);
+  const timestamp = new Date().toISOString();
+  const fileName = `${OUTPUT_FOLDER}/${dayString}/${hourString}/${timestamp}-${rand}.md`;
+
+  fs.writeFile(fileName, summary, (err) => {
+    if (err) {
+      console.error("Error saving summary to file:", err);
+    } else {
+      console.info(`Summary saved to ${fileName} successfully.`);
+    }
+  });
+}
+
+async function fetchNews() {
+  try {
+    const response = await axios.get(
+      `https://newsapi.org/v2/everything?q=cats&sortBy=publishedAt&pageSize=10&apiKey=${process.env.NEWS_API_KEY}`
+    );
+    return response.data;
+  } catch (error) {
+    console.error("Error fetching news:", error.message);
+
+    return null;
+  }
+}
+
+async function summarizeText(text) {
+  const openaiClient = new OpenAI({
+    apiKey: process.env.OPENAI_API_KEY,
+  });
+
+  try {
+    const gptResponse = await openaiClient.chat.completions.create({
+      messages: [
+        {
+          role: "system",
+          content: `
+          - You are the world's most advanced AI news analyst.
+          - Parse and summarise your reading of the article in less than 6 sentences.
+          - IMPORTANT: Use appropriate line breaks, punctuation, and grammar.
+          
+          - IMPORTANT: Format the summary in Markdown format.
+          - IMPORTANT: Use the following format for your summary:
+          ----
+          Example summary:
+          # Title of the summary
+          
+          This is the first sentence of the summary (paragraph)
+          \n (line break)
+
+          This is the second sentence of the summary (paragraph)
+          \n (line break)
+
+          This is the third sentence of the summary (paragraph)
+          \n (line break)
+
+          ... and so on.
+          ---
+      
+          - You are given the following news article html to parse summarize:
+          ${text}
+          `,
+        },
+      ],
+      model: "gpt-3.5-turbo",
+    });
+
+    const summary = gptResponse.choices[0].message.content;
+    return summary;
+  } catch (error) {
+    console.error("Error summarizing text:", error.message);
+
+    return "";
+  }
+}
+
+async function visitNewsArticle(url) {
+  try {
+    const response = await axios.get(url);
+
+    const $ = cheerio.load(response.data);
+    const article = $("article");
+    const articleTextContent = article.children("div").text();
+    const prettyArticle = pretty(articleTextContent);
+
+    return prettyArticle;
+  } catch (error) {
+    console.error("Error fetching news:", error.message);
+    return null;
+  }
+}
+
+async function runBot() {
```

Comment:
This is what you would consider the agent. It's a mix of if else statements and ai invocations.

Let's break it down

- We start by setting up the folders we'll need later on
- Then we set a while loop that never ends
- Inside this loop, we fetch the news json and if there is some data
- Create a variable that will hold the network requests to fetch the HTML for each news article
- We loop through each news article and
- - Visit the webpage for that news article but we don't await it so it doesn't resolve yet. Instead we push it to the variable we created above.
- Then we await all the network requests so that they resolve in parallel
- Then we filter these HTMLs to ensure we only process valid ones
- Then we do the same thing again and create another variable to store all the network requests we will create for summarising each page
- Loop through the valid news htmls and for each html
- - We create a network request to summarise the html with ai using `summarizeText`. You'll notice we attached a `.then` to this promise when we didn't for `visitNewsArticle`. This is because we want each individual summary to be saved immediately it's resolved. We don't want to do it in a separate step.
- - Finally, we push this promise to the variable created earlier
- Then we await these network requests to fetch the summaries and save them
- We're done for this loop
- The bot will now go to sleep for 12 hours only to repeat this process again.